must-call-checker/README.md

```diff
@@ -0,0 +1,118 @@
+by the `-AcheckMustCall` flag to the Object Construction Checker. You should use the Object
+Construction Checker directly rather than using the MustCall Checker alone.
+
+The MustCall Checker produces a conservative overapproximation of the set of methods that might
+actually need to be called on an object. The Object Construction Checker then conservatively assumes
+that this approximation (the `@MustCall` type) must be fulfilled.
+
+For example, an object whose type is `java.io.OutputStream` might
+have an obligation to call the `close()` method, to release an underlying file resource. Or,
+such an `OutputStream` might not have such an obligation, if the underlying resource is, for
+example, a byte array. Either of these dynamic types can be represented by the static type
+`@MustCall({"close"}) OutputStream`, which can be read as "an OutputStream that might need
+to call close before it is deallocated". The Object Construction Checker would enforce that the
+type of such an object in its hierarchy is a subtype of `@CalledMethods({"close"})` at each
+point it may become unreachable.
+
+### Explanation of qualifiers and type hierarchy
+
+The MustCall Checker supports two qualifiers: `@MustCall` and `@MustCallAny`. The `@MustCall`
+qualifier's arguments are the methods that the annotated type
+should call. The type `@MustCall({"a"}) Object obj` means
+that before `obj` is deallocated, `obj.a()` should be called.
+The type hierarchy is:
+
+                     @MustCallAny
+                           |
+                          ...
+                           |
+               @MustCall({"m1", "m2"})     ...
+                    /             \
+       @MustCall({"m1"})     @MustCall({"m2"})  ...
+                    \            /
+                    @MustCall({})
+
+Note that `@MustCall({"m1", "m2"}) Object` represents more possible values than
+`@MustCall({"m1"}) Object` does, because `@MustCall({"m1"}) Object` can only
+contain an object that needs to call `m1` or one that needs to call nothing - an
+object that needs to call `m2` (or both `m1` and `m2`) is not permitted.
+`@MustCall({"m1", "m2"}) Object` represents all objects that need to
+call `m1`, `m2`, both, or neither; it cannot not represent an object that needs
+to call some *other* method.
+
+The default type in unannotated code is `@MustCall({})`.  Given an expression of
+unannotated type, its value is not known to be required to call any methods.
+The user must provide a specification for any type whose must-call obligations
+they want to enforce.
+
+The programmer should rarely (if ever) write the top annotation (`@MustCallAny`), because
+the Object Construction Checker cannot verify it by matching with a corresponding `@CalledMethods`
+annotation. `@MustCallAny` theoretically represents a value with an unknowable or infinite set
+of must-call methods.
+
+### Example use
+
+If a user annotates a class declaration with a `@MustCall(...)` annotation, any use of the class's
```

Comment:
Please discuss the relationship between this syntax and `@DefaultQualifierForUse`.  I think it might be the same, but I'm not sure.

Reply:
This is a standard syntax supported by the Checker Framework; it's only used here because it makes the example clear. Adding an explanation, in my opinion, would distract from the core of the example. From the manual (https://checkerframework.org/manual/#default-for-use):

> If a type declaration is annotated with a qualifier bound, but not a `@DefaultQualifierForUse`, then the qualifier bound is added to all unannotated uses of that type (except locals). For example, `@C class MyClass` is equivalent to:
```
@DefaultQualifierForUse(C.class)
@C class MyClass {}
```

Reply:
Is there a specific suggestion you have on how to clarify this example that won't distract from its main point, which is to show users how to actually place an `@MustCall` annotation?

Reply:
Note to self, mention that this is a standard syntax and link that manual section, nothing else